src/tools/tests/schema2kotlin-test.ts

```diff
@@ -232,27 +263,181 @@ describe('schema2kotlin', () => {
+      `@Suppress("UNCHECKED_CAST")
+    class Thing(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("Thing", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates variable entity with private constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads ~a with {num: Number}
+         `,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1 private constructor(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : VariableEntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates copy and mutate by entity fields', async () => await assertCopyMethods(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = Thing(num = num)
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = Thing(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        )`
+    ));
+    // TODO(alxr): Why do we omit the docstring for Wasm?
+    it('generates only copy method by entity fields for wasm', async () => await assertCopyMethodsForWasm(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `
+        fun copy(num: Double = this.num) = Thing(num = num)
+        `
+    ));
+    it('copies underlying data for type variables.', async () => await assertCopyMethods(
+      `particle T
+         h1: reads ~a with Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = Thing(num = num)
+            .also { this.copyInto(it) }
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = Thing(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        ).also { this.copyInto(it) }`
```

Comment:
To the reader it may be not obvious that `copyInto` is defined on the VariableEntityBase (as opposed to EntityBase) and operates only on raw values, instead of the deserialized ones. If you don't have this background and you read this code it seems like copyInto could overwrite the values you pass in the constructor.

If I were you I would rename `copyInto` to something like `copyRawValuesInto`